Tilemap refactor

jthomassie opened the pull request:
Includes several 'low-hanging fruit' updates:
- changed map min zoom to 1 and max zoom to 18, closes #3160
- simplified code by removing a few methods
- eliminated redraw of circle markers on map zoom by using distance instead of pixels
- improved sizing of circles by using geohash grid min distance of rect, closes #2370, closes #3159 
- added a couple of map types to the view options: geohash grid rectangles, pins, closes #3158
- added shading to the default scaled circles in order to add a legend for this type
- added tests, closes #1917

rashidkpc commented:
Can you add a detailed description of each bug fix and enhancement and also link this back to the issues with "closes #blah"

rashidkpc commented on src/kibana/components/vislib/visualizations/tile_map.js:
```diff
@@ -404,25 +512,26 @@ define(function (require) {
+      var ne = L.latLng([gh[2][1], gh[2][0]]);
+      var sw = L.latLng([gh[0][1], gh[0][0]]);
+      var ew = Math.floor(nw.distanceTo(ne));
+      var ns = Math.floor(nw.distanceTo(sw));
```
two letter variables are hard to read, can you expand these to be descriptive of their use?

w33ble commented:
There's an error as soon as I create a tilemap vis, on tile_map.js, line 121 - looks like a missing DOM element?

![screenshot 2015-03-25 16 30 17](https://cloud.githubusercontent.com/assets/404731/6837664/303f2a3c-d30c-11e4-8e5f-08b087b4bc7f.png)

w33ble commented on src/kibana/components/vislib/visualizations/tile_map.js:
```diff
@@ -47,29 +42,38 @@ define(function (require) {
 
       // clean up old maps
-      _.invoke(self.maps, 'destroy');
+      _.invoke(self.maps, self.destroy);
```
This will iterate over every map and call `destroy` - which itself iterates over every map to destroy it. You don't need `invoke` or iteration here - `self.destroy()` should be enough.

w33ble commented on src/kibana/components/vislib/visualizations/tile_map.js:
```diff
@@ -47,29 +42,38 @@ define(function (require) {
+
       var self = this;
       var $elem = $(this.chartEl);
       var div;
```
`div` is never used in this scope, and it's clobbered in the `each` call (line 74). It should be inside that iteration.

w33ble commented on src/kibana/components/vislib/visualizations/tile_map.js:
```diff
@@ -47,29 +42,38 @@ define(function (require) {
           }
-          if (self._attr.lastCenter) {
-            mapCenter = self._attr.lastCenter;
+          if (self._attr.mapCenter) {
```
Is this a typo? The same thing happens regardless of the conditional. Also, the map center point and zoom are lost completely in this PR - is this why?

w33ble commented:
In this PR, the state of the map is lost with every change, which makes it basically unusable.

Master (move/zoom the top map, apply changes, the bottom map changes too - this seems wrong, but is workable):

![2015-03-25 17_24_41](https://cloud.githubusercontent.com/assets/404731/6838286/d036fe5a-d313-11e4-9c70-40ee31262410.gif)

PR 3445 (move/zoom, apply changes, both maps are completely reset):

![2015-03-25 17_27_06](https://cloud.githubusercontent.com/assets/404731/6838296/1210cf7c-d314-11e4-8814-03d4feb95268.gif)

w33ble commented:
Shaded geohash grid doesn't appear to work correctly

![screenshot 2015-03-25 17 30 24](https://cloud.githubusercontent.com/assets/404731/6838316/7f42c82a-d314-11e4-88fc-240bb4b554f7.png)

jthomassie commented:
@w33ble @rashidkpc I have updated this with several commits to address your comments. 
Ready for review.

w33ble commented:
![2015-04-02 16_54_05](https://cloud.githubusercontent.com/assets/404731/6976030/a8e1bb56-d958-11e4-8b5b-65a3271dfb15.gif)

Changing the marker type resets the map state. It doesn't center or anything even - not sure how it's picking the new map center.

w33ble commented:
I see how it's choosing the center - that's just the leaflet center point.

Maps don't see to do an initial "fit to" call now. The top is after I clicked that button, the bottom is how it loaded.

![screenshot 2015-04-02 17 01 23](https://cloud.githubusercontent.com/assets/404731/6976111/c7df90e0-d959-11e4-9d44-ca7c77c16380.png)

jthomassie commented:
@w33ble misunderstood your previous example. This behaves the same as master now.

w33ble commented on src/kibana/components/vislib/visualizations/tile_map.js:
```diff
@@ -208,36 +274,81 @@ define(function (require) {
-          return L.circleMarker(latlng, {
-            radius: rad
-          });
+          var gh = feature.properties.rectangle;
```
`gh` is a pretty bad name, I have no idea what it's short for. Also, a note about what's going on here would be helpful, in particular, why [0,0], [0,1], [2,1], [2,0].

w33ble commented on src/kibana/components/vislib/visualizations/tile_map.js:
```diff
@@ -208,36 +274,81 @@ define(function (require) {
+          layer.on({
+            mouseover: function (e) {
+              var layer = e.target;
+              if (!L.Browser.ie && !L.Browser.opera) {
```
Why is this exception here? A comment would be helpful.

w33ble commented on src/kibana/components/vislib/visualizations/tile_map.js:
```diff
@@ -404,25 +479,29 @@ define(function (require) {
+      var centerPoint = feature.properties.center;
+      var geohashRect = feature.properties.rectangle;
+
+      var center = L.latLng([centerPoint[1], centerPoint[0]]);
```
A comment about the coords here would be helpful ([2][0], [3][1], etc)

w33ble commented on src/kibana/components/vislib/visualizations/tile_map.js:
```diff
@@ -433,90 +512,15 @@ define(function (require) {
      */
-    TileMap.prototype.radiusScale = function (count, max, precision) {
-      // exp = 0.5 for true square root ratio
+    TileMap.prototype.radiusScale = function (count, max, precision, feature) {
```
`precision` isn't being used anymore, it shouldn't be passed in.

w33ble commented on src/kibana/components/vislib/visualizations/tile_map.js:
```diff
@@ -527,19 +531,18 @@ define(function (require) {
+      var cScale = this._attr.cScale = d3.scale.quantize()
         .domain([min, max])
         .range(colors);
       if (max === min) {
```
Can you fix the indenting above here - the lines should be flush with the `var cScale` line. Also, please add a blank line before the `if`

w33ble commented:
Aside from the code syntax comments, this seems functional and looks good to me. Passing to @panda01 just to get another set of eyes on it, since there's a lot here to review.

panda01 commented on src/kibana/components/vislib/visualizations/tile_map.js:
```diff
@@ -125,30 +118,104 @@ define(function (require) {
+
+      // super min and max from all chart data
+      var min = mapData.properties.allmin;
+      var max = mapData.properties.allmax;
```
Hmm, it doesn't appear to me like these variables are used.

jthomassie commented:
@w33ble @panda01thanks for the review and feedback. I know there it had too many commits for one PR.  I have updated to address all of your comments:
- Added comments to some of the code
- Renamed a two-letter var
- Deleted unused params and vars from previous versions
- Fixed a test that failed when an unused param was deleted

panda01 commented on src/kibana/components/vislib/visualizations/tile_map.js:
```diff
@@ -125,30 +118,100 @@ define(function (require) {
+      var self = this;
+
+      var length = mapData.properties.length;
+      var precision = mapData.properties.precision;
```
It appears to me that these variables aren't used in this function.

jthomassie replied:
@panda01 thanks. Fixed this one.

panda01 commented on src/kibana/components/vislib/visualizations/tile_map.js:
```diff
@@ -156,47 +219,46 @@ define(function (require) {
+      var max = mapData.properties.allmax;
+
       var length = mapData.properties.length;
       var precision = mapData.properties.precision;
```
It seems to me these variables aren't used any longer.

jthomassie replied:
@panda01 min and max are used in this method

panda01 replied:
Not the min and max, the length and precision vars

panda01 commented on src/kibana/components/vislib/visualizations/tile_map.js:
```diff
@@ -208,36 +270,87 @@ define(function (require) {
+
       var featureLayer = L.geoJson(mapData, {
         pointToLayer: function (feature, latlng) {
           var count = feature.properties.count;
```
It seems like this variable isn't used.

jthomassie replied:
@panda01 bounds is used here

panda01 replied:
The var `count` is the variable I'm referring to here.

panda01 commented on src/kibana/components/vislib/visualizations/tile_map.js:
```diff
@@ -208,36 +270,87 @@ define(function (require) {
+          });
         },
         style: function (feature) {
           var count = feature.properties.count;
```
This appears to be the same function as [https://github.com/elastic/kibana/pull/3445/files#diff-220771d5f556b3a25ecdffc1771b40afR292]().

Perhaps we can create one function to be shared between the two?

jthomassie replied:
@panda01 Yes, will add a method to apply this style which is used in three map types. Thanks!
